crates/ecs/src/systems/command_buffers.rs

```diff
@@ -0,0 +1,715 @@
+    }
+
+    #[test]
+    fn system_removing_same_component_type_from_entity_second_time_is_a_noop() {
+        let twice_removal_system = |entity: Entity, commands: Commands| {
+            commands.remove_component::<D>(entity);
+            commands.remove_component::<D>(entity);
+        };
+
+        let (app, _, _, _) = set_up_app_with_systems_and_entities([twice_removal_system]);
+
+        let mut runner = app.into_tickable::<Sequential, Unordered>().unwrap();
+        runner.tick().unwrap();
+        runner.playback_commands().unwrap();
+
+        let component_values: Vec<_> = read_component_values::<D>(&runner);
+
+        assert_eq!(
+            component_values.len(),
+            0,
+            "all components of type D should have been removed"
+        );
+    }
+
+    #[test]
+    fn two_systems_removing_same_component_type_from_same_entity_removes_that_component_type_once()
+    {
+        let removing_system0 = |entity: Entity, commands: Commands| {
+            commands.remove_component::<D>(entity);
+        };
+        let removing_system1 = |entity: Entity, commands: Commands| {
+            commands.remove_component::<D>(entity);
+        };
+
+        let (app, _, _, _) =
+            set_up_app_with_systems_and_entities([removing_system0, removing_system1]);
+
+        let mut runner = app.into_tickable::<Sequential, Unordered>().unwrap();
+        runner.tick().unwrap();
+        runner.playback_commands().unwrap();
+
+        let component_values: Vec<_> = read_component_values::<D>(&runner);
+
+        assert_eq!(
+            component_values.len(),
+            0,
+            "all components of type D should have been removed"
+        );
+    }
+
+    #[test]
+    fn system_can_create_new_entity_with_multiple_components() {
+        let creation_system = |entity: Entity, commands: Commands| {
+            let creation = EntityCreation::default()
+                .with_component(A(entity.id as i32))
+                .with_component(B("hi".to_owned()))
+                .with_component(C(1.0));
+            commands.create(creation);
+        };
+
+        let (app, _, _, _) = set_up_app_with_systems_and_entities([creation_system]);
+
+        let mut runner = app.into_tickable::<Sequential, Unordered>().unwrap();
+        runner.tick().unwrap();
+        runner.playback_commands().unwrap();
+
+        let a_values: Vec<_> = read_component_values::<A>(&runner);
+        let b_values: Vec<_> = read_component_values::<B>(&runner);
+        let c_values: Vec<_> = read_component_values::<C>(&runner);
+
+        assert_eq!(
+            runner.world.entities.len(),
+            6,
+            "three new entities should have been created"
+        );
+        assert_eq!(
+            a_values.len(),
+            3,
+            "3 entities with components A, B and C should have been created"
+        );
+        assert_eq!(a_values.len(), b_values.len());
+        assert_eq!(b_values.len(), c_values.len());
+    }
```

Comment:
Found a problem that causes a panic if you add and remove entities repeatedly. Here is a simple test for that:
```suggestion
    }
    
    #[test]
    fn creating_and_removing_entities_multiple_times_does_not_cause_panic() {
        let removing_creation_system = |entity: Entity, commands: Commands| {
            commands.remove(entity);
            let creation = EntityCreation::default().with_component(C(1.0));
            commands.create(creation);
        };

        let (app, _, _, _) = set_up_app_with_systems_and_entities([removing_creation_system]);

        let mut runner = app.into_tickable::<Sequential, Unordered>().unwrap();

        for _ in 0..10 {
            runner.tick().unwrap();
            runner.playback_commands().unwrap();
        }
    }
```

Reply:
Thanks, I'll fix that!